### PR TITLE
NMS-12414: Added BMP telemetry adapter

### PR DESCRIPTION
**Pull request by fooker**
The new telemetryd adapter collects metrivs from BMP statisic messages
and persist them as OpenNMS metrices associated with the router node.

Each peer of the router will be a subresource with the associated
statistics.

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12414

### CODE REVIEW

File: features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapter.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2017-2017 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2017 The OpenNMS Group, Inc.
```
Comment (mvrueden) [inline features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapter.java]: Wrong copyright date header. Should be `2019-2019` and `1999-2019`.

File: features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapterFactory.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2017-2017 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2017 The OpenNMS Group, Inc.
```
Comment (mvrueden) [inline features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapterFactory.java]: Wrong dates in header.

File: features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapter.java (hunk 1 above)
Comment (mvrueden) [inline features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapter.java]: header. Should be 2019-2019 and 1999-2019.

File: features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapterFactory.java (hunk 2 above)
Comment (mvrueden) [inline features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapterFactory.java]: Wrong dates in header.

File: features/telemetry/protocols/collection/src/main/java/org/opennms/netmgt/telemetry/protocols/collection/AbstractPersistingAdapter.java [hunk 3]
```diff
@@ -1,8 +1,8 @@
 /*******************************************************************************
  * This file is part of OpenNMS(R).
  *
- * Copyright (C) 2017-2017 The OpenNMS Group, Inc.
- * OpenNMS(R) is Copyright (C) 1999-2017 The OpenNMS Group, Inc.
+ * Copyright (C) 2019 The OpenNMS Group, Inc.
```
Comment (mvrueden) [inline features/telemetry/protocols/collection/src/main/java/org/opennms/netmgt/telemetry/protocols/collection/AbstractPersistingAdapter.java]: If the file was created before and updated, the copyright headers should state `<createYear>-<currentYear>`, in this case `2017-2019`.

File: features/telemetry/protocols/collection/src/main/java/org/opennms/netmgt/telemetry/protocols/collection/AbstractScriptedPersistingAdapter.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2017-2017 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2017 The OpenNMS Group, Inc.
```
Comment (mvrueden) [inline features/telemetry/protocols/collection/src/main/java/org/opennms/netmgt/telemetry/protocols/collection/AbstractScriptedPersistingAdapter.java]: Copy right header dates.

File: features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapter.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        // Extract peer details
+        final String peerAddress = getString(document, "peer", "address").get();
+        final Instant timestamp = Instant.ofEpochSecond(getInt64(document, "peer", "timestamp", "epoch").get(),
+                                                        getInt64(document, "peer", "timestamp", "nanos").orElse(0L));
+        final String as = Long.toString(getInt64(document, "peer", "as").get());
+        final String id = Long.toString(getInt64(document, "peer", "id").get());
+
+        // Build resource for the peer
+        final NodeLevelResource nodeResource = new NodeLevelResource(agent.getNodeId());
+        final GenericTypeResource peerResource = new GenericTypeResource(nodeResource, RESOURCE_TYPE, peerAddress);
```
Comment (j-white) [inline features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapter.java]: Can use a `DeferredGenericTypeResource` instead so we don't need to define it in code. The resource type should be added to the config in `resource-types.d`.

Comment (fooker) [reply features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapter.java]: Didn't know about that. Changed.

**Pull request merged**